### PR TITLE
Fix batch_norm dtype mismatch on MPS (fixes #154887)

### LINKED ISSUE #154887: MPS batch_norm mixed dtype failure
### 🐛 Describe the bug

```python
import torch

torch.nn.functional.batch_norm(
    torch.rand((2, 3, 4), dtype=torch.float16, device="mps"),
    torch.rand((3,), dtype=torch.float32, device="mps"),
    torch.rand((3,), dtype=torch.float32, device="mps"))
```

Alternatively, you can also trigger this error calling `_native_batch_norm_legit_no_training`
```python
torch.ops.aten._native_batch_norm_legit_no_training.default(
    torch.rand((2, 3, 4), dtype=torch.float16, device="mps"),
    None,
    None,
    torch.rand((3,), dtype=torch.float32, device="mps"),
    torch.rand((3,), dtype=torch.float32, device="mps"),
    0.1,
    1e-05)
```

Error message:
```
loc("mps_normalization"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/79ef05cb-ffe7-11ef-b000-f2a857e00a32/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":45:0)): error: output type 'tensor<2x3x4xf16>' and mean type 'tensor<1x3x1xf32>' are not broadcast compatible
LLVM ERROR: Failed to infer result type(s).
zsh: abort
```

### Versions

torch==2.7.0

### PR DESCRIPTION
**Pull request by roei-shlezinger**
This PR fixes a dtype mismatch in MPS batch_norm where float16 inputs and float32 running stats caused MPSGraph to fail during type inference.
The fix adds a small dtype alignment step (casting mean/var/weight/bias to the input dtype when needed).
No functional or numerical changes-only dtype consistency before calling normalizationWithTensor.

Fixes #154887

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Normalization.mm [hunk 1]
```diff
@@ -273,11 +273,23 @@ Check if running mean exists (maybe do this check before making graph)
       }
 
       // Compute output of batch norm
+      auto castTensorIfNeeded = [&](MPSGraphTensor* tensor, NSString* name) -> MPSGraphTensor* {
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/Normalization.mm]: Instead of this lambda, probably should just use `castMPSTensor` from `aten/src/ATen/native/mps/OperationUtils.h`

File: test/test_mps.py [hunk 2]
```diff
@@ -1831,6 +1831,28 @@ def test_batch_norm_backward_weight_bias_gradients(self):
...
+    def test_batch_norm_mixed_dtypes(self):
+        # See issue: https://github.com/pytorch/pytorch/issues/154887#issue-3111485538
+        input_mps = torch.rand((2, 3, 4), dtype=torch.float16, device="mps")
+        input_cpu = input_mps.clone().cpu()
+        mean_mps = torch.rand((3,), dtype=torch.float32, device="mps")
+        mean_cpu = mean_mps.clone().cpu()
+        var_mps = torch.rand((3,), dtype=torch.float32, device="mps")
+        var_cpu = var_mps.clone().cpu()
+
+        cases = [
```
Comment (kurtamohler) [inline test/test_mps.py]: nit: Might as well `@parametrize` these

File: test/test_mps.py [hunk 3]
```diff
@@ -1831,6 +1831,28 @@ def test_batch_norm_backward_weight_bias_gradients(self):
         self.assertEqual(bn_cpu.weight.grad, bn_mps.weight.grad, atol=1e-5, rtol=1e-5)
         self.assertEqual(bn_cpu.bias.grad, bn_mps.bias.grad, atol=1e-5, rtol=1e-5)
 
+    def test_batch_norm_mixed_dtypes(self):
+        # See issue: https://github.com/pytorch/pytorch/issues/154887#issue-3111485538
+        input_mps = torch.rand((2, 3, 4), dtype=torch.float16, device="mps")
```
Comment (kurtamohler) [inline test/test_mps.py]: Looks like your change will enable any `dtype_a` -> `dtype_b` cast, not just float16 -> float32, so we should probably test all of the supported cases, using `@parametrize`.

Also, I think the CPU impl does not support all cases and raises an error instead, so MPS should probably error in the same cases

File: test/test_mps.py [hunk 4]
```diff
@@ -1831,6 +1831,28 @@ def test_batch_norm_backward_weight_bias_gradients(self):
...
+        input_cpu = input_mps.clone().cpu()
```
Comment (kurtamohler) [inline test/test_mps.py]: nit: no need to clone, since `.cpu()` clones anyway

**Review by kurtamohler — COMMENTED** [review]
Thanks for the fix, just a few suggestions

File: test/test_mps.py [hunk 5]
```diff
@@ -1831,6 +1831,40 @@ def test_batch_norm_backward_weight_bias_gradients(self):
...
+    def test_batch_norm_mixed_dtypes(self, case_name, fn, input_dtype, param_dtype, expect_error):
+        # See issue: https://github.com/pytorch/pytorch/issues/154887
+        input_mps = torch.rand((2, 3, 4), dtype=input_dtype, device="mps")
+        input_cpu = input_mps.cpu()
+        mean_mps = torch.rand((3,), dtype=param_dtype, device="mps")
+        mean_cpu = mean_mps.cpu()
+        var_mps = torch.rand((3,), dtype=param_dtype, device="mps")
+        var_cpu = var_mps.cpu()
+
+        if expect_error:
```
Comment (kurtamohler) [inline test/test_mps.py]: nit: Rather than parameterizing `expect_error`, it would be better to run the CPU op in a try/except first to check if it raises an error. That way if more dtype combos are supported in the future, the test doesn't have to be updated

File: test/test_mps.py [hunk 6]
```diff
@@ -1831,6 +1831,40 @@ def test_batch_norm_backward_weight_bias_gradients(self):
         self.assertEqual(bn_cpu.weight.grad, bn_mps.weight.grad, atol=1e-5, rtol=1e-5)
         self.assertEqual(bn_cpu.bias.grad, bn_mps.bias.grad, atol=1e-5, rtol=1e-5)
 
+    @parametrize("case_name,fn", [
+        ("functional", lambda inp, mean, var: torch.nn.functional.batch_norm(
+            inp, mean, var, training=False, eps=1e-5)),
+        ("native_legit_no_training", lambda inp, mean, var: torch.ops.aten._native_batch_norm_legit_no_training.default(
+            inp, None, None, mean, var, 0.1, 1e-5)[0]),
+    ])
+    @parametrize("input_dtype,param_dtype,expect_error", [
```
Comment (kurtamohler) [inline test/test_mps.py]: nit: Better to separate the dtype params so you don't have to write out all the different combos

```
@parametrize("input_dtype", [...])
@parametrize("param_dtype", [...])
```

**Review by kurtamohler — APPROVED** [review]
Just a few nits but it looks good to me.

I think the `check_mixed_data_type` oddity is ok. It would be good to fix its error message to be device-agnostic and move it to a more common place than `ATen/native/cpu`, but that should be done in a separate PR

**Review by malfet — APPROVED** [review]
